src/Microsoft.AspNetCore.Blazor.Server/Circuits/RenderBatchFormatterResolver.cs

```diff
@@ -0,0 +1,44 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using MessagePack;
+using MessagePack.Formatters;
+using Microsoft.AspNetCore.Blazor.Rendering;
+using System;
+using System.IO;
+
+namespace Microsoft.AspNetCore.Blazor.Server.Circuits
+{
+    /// <summary>
+    /// A MessagePack IFormatterResolver that provides an efficient binary serialization
+    /// of <see cref="RenderBatch"/>. The client-side code knows how to walk through this
+    /// binary representation directly, without it first being parsed as an object graph.
+    /// </summary>
+    internal class RenderBatchFormatterResolver : IFormatterResolver
+    {
+        public IMessagePackFormatter<T> GetFormatter<T>()
+            => typeof(T) == typeof(RenderBatch) ? (IMessagePackFormatter<T>)RenderBatchFormatter.Instance : null;
+
+        private class RenderBatchFormatter : IMessagePackFormatter<RenderBatch>
+        {
+            public static readonly RenderBatchFormatter Instance = new RenderBatchFormatter();
+
+            // No need to accept incoming RenderBatch
+            public RenderBatch Deserialize(byte[] bytes, int offset, IFormatterResolver formatterResolver, out int readSize)
+                => throw new NotImplementedException();
+
+            public int Serialize(ref byte[] bytes, int offset, RenderBatch value, IFormatterResolver formatterResolver)
```

Comment:
I toyed with using some kind of object pool for the underlying buffer here to avoid initializing multi-KB arrays on every interaction. Didn't do it at first because I was just focused on the functionality rather than that sort of perf optimization.

What do you think - is the cost of large object allocation big enough to warrant putting in an object pool here? Or do we decide later?

Reply:
It will be worth it if it's LOH

Reply:
I'm tracking this as something to log an issue for follow up. Let me know if you think it needs to be addressed immediately.

Reply:
Tracking it as a future enhancement sounds fine to me.

Reply:
Logged #1132